lib/pure/asyncdispatch.nim

```diff
@@ -1851,6 +1909,105 @@ proc send*(socket: AsyncFD, data: string,
+  vd.virtualMuxHandle.unregister()
+  vd.virtualMuxHandle.close()
+  vd.virtualMuxHandle = nil
+  vd.nextVirtualHandle = 0.VirtualFD
+
+proc unregister*(ev: VirtualAsyncEvent) =
+  ## Unregisters event ``ev``.
+  doAssert(ev.vFD != InvalidVirtualFD, "Event is not registered in the queue!")
+
+  let vd = ev.vd
+  withLockIfThreads ev.eventLock:
+    ev.vd.virtualHandles.del ev.vFD
+    ev.vd = nil
+    ev.cb = nil
+    ev.vFD = InvalidVirtualFD
+
+  if vd.virtualHandles.len == 0:
+    # lazy close the physical event with the Dispatcher
+    # The main reason we need this is to make
+    # hasPendingOperations still work
+    # without modifying the ioselector code.
+    vd.close
+
+proc initVirtualEventDispatcher(p: PDispatcher) =
+  p.vd.virtualMuxHandle = newAsyncEvent()
+
+  proc virtualMuxCB(fd: AsyncFD): bool {.closure,gcsafe.} =
+    # find the virtual events that triggered the physical event and
+    # add them to the callback list.
+    # Not very efficient, but requires the least coordination between threads.
+    
+    proc vcbFactory(vfd: VirtualFD) : proc() {.closure, gcsafe.} =
+      let ev = p.vd.virtualHandles[vfd]
+      result = proc() {.closure,gcSafe.} =
+        if ev.cb(vfd.AsyncFD):
+          # the convention is that if the callback returns true, we unregister the event.
+          ev.unregister
+
+    for vfd, ev in p.vd.virtualHandles:
+      withLockIfThreads ev.eventLock:
+        if ev.triggered:
+          ev.triggered = false
+          p.callbacks.addLast(vcbFactory(vfd))
+
+    # always return false b/c we never want to unregister this event
+    return false
+
+  p.vd.virtualMuxHandle.addEvent(virtualMuxCB)
+
+proc newVirtualAsyncEvent*(): VirtualAsyncEvent =
+  ## Creates a new thread-safe ``VirtualAsyncEvent`` object.
+  ##
+  ## New ``VirtualAsyncEvent`` object is not automatically registered with
+  ## dispatcher like ``AsyncSocket``.
+  result = cast[VirtualAsyncEvent](allocShared0(sizeof(VirtualAsyncEventImpl)))
+  result.cb = nil
+  result.vFD = InvalidVirtualFD
+  when compileOption("threads"):
+    initRLock result.eventLock
+
+proc trigger*(ev: VirtualAsyncEvent) =
+  ## Sets new ``VirtualAsyncEvent`` to signaled state.
+  withLockIfThreads ev.eventLock:
+    if ev.vFD == InvalidVirtualFD:
+      # triggering on an event that is not registered is a noop.
+      return
+    ev.triggered = true
+
+    # send the signal to wake up the dispatcher thread.
+    trigger(ev.vd.virtualMuxHandle)
+
+proc addEvent*(ev: VirtualAsyncEvent, cb: Callback) =
+  ## Registers callback ``cb`` to be called when ``ev`` will be signaled
+  doAssert(ev.vFD == InvalidVirtualFD, "Event is already registered in the queue!")
+
+  let p = getGlobalDispatcher()
+
+  if p.vd.virtualMuxHandle == nil:
+    # lazy register the physical event with the Dispatcher
+    # The main reason we need this is to make hasPendingOperations still work
+    # without modifying the ioselector code.
```

Comment:
can you elaborate on this? Don't really understand why this lazy loading is necessary.

Reply:
It has to do with `hasPendingOperations()` and the multiplexer `io select` object.

The simplest approach is to simply create a singleton `AsyncEvent` object attached to the `VirtualEventDispatcher` , which you initialize on creation.
This `AsyncEvent` essentially lives forever and waits for `VirtualEvents` to trigger it.

This has one big problem. The *lives forever* part. The `io selector` is never closed. It must remain alive. This means the underlying ioselector implementation will always have at least one open `io select` that it is listening on.

If any code relies on all `io selects` being closed before continuing, that code will never be able to run. `hasPendingOperations()` and any code that relies on it is a prime example of code like this.

In #12232 I *"fixed"* this problem by modifying all of the io_selector implementations and `asyncdispatch.nim` to treat a single open `io select` as the same as *0* open `io selects` (no pending events).

After thinking about it, I realized that has approach has big problems:

- First, it is a very dramatic and invasive change to the ioselector implementations. That code is foundational, so I want to be very careful with changes that I make there. The changes in #12232 seemed too big, and not safe.

- Second, It's possible that code might rely on the process actually having all physical kernel `io selectors` closed (for a graceful shutdown or resource management for example). If we treat 1 open `io selector` the same as 0, we are essentially lying to the user, and that is not ok IMO.

I don't love lazy loading, because it is non-deterministic, but in this case, I think it's the right tool for the job. This approach is a modular solution to all of the above problems:

- It works with 0 changes to the underyling `ioselector` code
- It works with 0 changes to the core asyncdispatch code.
    - By core I mean no changes to `AsyncEvent` or any core run loop code such as `runOnce()`
- `hasPendingOperations()` works correctly **without** lying to the user
- All changes are contained within the `VirtualEvent` additions.

All of that adds up to make this a much safer PR.